server/src/com/thoughtworks/go/server/security/BasicAuthenticationFilter.java

```diff
@@ -32,18 +39,73 @@ protected Boolean initialValue() {
             return false;
         }
     };
+    private static final Logger LOG = Logger.getLogger(BasicAuthenticationFilter.class);
 
     @Override
     public void doFilterHttp(HttpServletRequest httpRequest, HttpServletResponse httpResponse, FilterChain chain) throws IOException, ServletException {
         try {
             isProcessingBasicAuth.set(true);
             super.doFilterHttp(httpRequest, httpResponse, chain);
+        } catch (Exception e) {
+            handleException(httpRequest, httpResponse, e);
         } finally {
             isProcessingBasicAuth.set(false);
         }
     }
 
+    public void handleException(HttpServletRequest httpRequest, HttpServletResponse httpResponse, Exception e) throws IOException {
+        String message = "There was an error authenticating you. Please check the server logs, or contact your the go administrator.";
+        if (hasAccept(httpRequest, "text/html") || hasAccept(httpRequest, "application/xhtml")) {
+            SecurityContext context = SecurityContextHolder.getContext();
+            httpRequest.getSession().setAttribute(AbstractProcessingFilter.SPRING_SECURITY_LAST_EXCEPTION_KEY, new RuntimeException(message));
+            httpRequest.setAttribute(SessionDenialAwareAuthenticationProcessingFilterEntryPoint.SESSION_DENIED, true);
+            context.setAuthentication(null);
```

Comment:
this causes an already logged in user to be logged out (if he tries to login with invalid credentials). intentional?